src/v/pandaproxy/json/requests/partition_offsets.h

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "json/json.h"
+#include "model/fundamental.h"
+#include "pandaproxy/json/types.h"
+#include "seastarx.h"
+
+#include <seastar/core/sstring.hh>
+
+#include <rapidjson/encodings.h>
+
+namespace pandaproxy::json {
+
+struct topic_partition_offset {
+    model::topic topic;
+    model::partition_id partition;
+    model::offset offset;
+    bool operator<(const topic_partition_offset& other) const {
+        return std::tie(topic, partition)
+               < std::tie(other.topic, other.partition);
+    };
+};
+
+template<typename Encoding = rapidjson::UTF8<>>
+class partition_offsets_request_handler {
+private:
+    enum class state {
+        empty = 0,
+        partitions,
+        tpo,
+        topic,
+        partition,
+        offset,
+    };
+
+    serialization_format _fmt = serialization_format::none;
+    state state = state::empty;
+
+public:
+    using Ch = typename Encoding::Ch;
+    using rjson_parse_result = std::vector<topic_partition_offset>;
+    rjson_parse_result result;
+
+    bool Null() { return false; }
+    bool Bool(bool) { return false; }
+    bool Int64(int64_t) { return false; }
+    bool Uint64(uint64_t) { return false; }
+    bool Double(double) { return false; }
+    bool RawNumber(const Ch*, rapidjson::SizeType, bool) { return false; }
+
+    bool Int(int i) {
+        switch (state) {
+        case state::partition:
+            result.back().partition = model::partition_id(i);
+            state = state::tpo;
+            return true;
+        case state::offset:
+            result.back().offset = model::offset(i);
+            state = state::tpo;
+            return true;
+        default:
+            return false;
+        }
+    }
+
+    bool Uint(unsigned u) {
+        switch (state) {
+        case state::partition:
+            result.back().partition = model::partition_id(u);
+            state = state::tpo;
+            return true;
+        case state::offset:
+            result.back().offset = model::offset(u);
+            state = state::tpo;
```

Comment:
question: why does each case end up in the same next tpo state?

Reply:
> question: why does each case end up in the same next tpo state?

@dotnwat Semantically, we're parsing an object of simple fields, the field order isn't known, receiving the key tells us which field we're on, receiving a value (as here), tells us the field is done, we're going back to parsing the object. 

Syntactically, I could `[[fallthrough]]`

Reply:
no need to change anything. was more of a question of how rapidjson works. thanks for explanation